### PR TITLE
[core] Implement TileCover for polygonal regions

### PR DESCRIPTION
**Pull request by asheemmamoowala**
The existing scanline-based tile cover implementation blocks when computing large rectangles. When extended to polygonal regions, this would further impact the performance and block interactivity in end-user applications.

This PR implements a streaming tile cover algorithm. An edge table-like data structure is created for all geometry, and then a modified scan-line is used for each tile row. Each row requires just one scan-line 
while accumulating the x-extents for each edge within that row. 

Usage:
```
    mbgl::util::TileCover tc(LatLngBounds::world(), zoom);
    while(tc.hasNext()) {
        UnwrappedtileID tile = *tc.next();
        ...
    };
```

- [x] Unit tests for polygons with holes and multi-polygons
- [x] A wrapper method `util::tileCover` that accepts arbitrary geometry and returns tiles.  
- [x] Benchmark tests for tile cover

cc @ivovandongen @mourner

### CODE REVIEW

File: include/mbgl/util/projection.hpp [hunk 1]
```diff
@@ -78,8 +78,9 @@ class Projection {
         return project_(latLng, worldSize(scale));
     }
 
-    static Point<double> project(const LatLng& latLng, uint8_t zoom) {
-        return project_(latLng, std::pow(2.0, zoom));
+    //Returns point on tile
+    static Point<double> project(const LatLng& latLng, int32_t zoom) {
```
Comment (asheemmamoowala) [inline include/mbgl/util/projection.hpp]: This method projects to a tile fraction and not to a tile-size based world. I'm not sure if it should be called `project` or exposed from `TileCoordinate`

File: test/util/tile_cover.test.cpp [hunk 2]
```diff
@@ -85,11 +85,167 @@ TEST(TileCover, SanFranciscoZ0Wrapped) {
...
+            {22.587890625,6.751896464843375},
+            {21.665039062499996,12.597454504832017},
+            {20.9619140625,8.189742344383703},
+            {18.193359375,14.3069694978258},
+            {16.611328125,8.667918002363134}
+        }
+    };
+
+    //These are all different from the js tile-cover lib
+    EXPECT_EQ(1742u, util::tileCover(spikyGeom, 10).size());
```
Comment (asheemmamoowala) [inline test/util/tile_cover.test.cpp]: I suspect that this test has different results based on floating point precision difference between the JS and CPP implementations. It's hard to reproduce with a small test case.

File: src/mbgl/util/tile_cover_impl.cpp [hunk 3]
```diff
@@ -234,47 +235,48 @@ TileCoverImpl::TileCoverImpl(int32_t z, const Geometry<double>& geom, bool proje
...
-    while (p != xps.end()) {
-        auto x_min = p->first;
-        // For closed geometries, consider odd even pairs of tile ranges
-        if (isClosed) { p++; }
-        auto x_max = p++->second;
-        scanCover(x_min, x_max, currentY);
+
+    auto x_min = xps[0].x0;
+    auto x_max = xps[0].x1;
+    int8_t nzRule = xps[0].winding ? 1 : -1;
```
Comment (mourner) [inline src/mbgl/util/tile_cover_impl.cpp]: `nzRule` can overflow if the scan line intersects a ton of line strings, right? Maybe worth bumping `int8_t` to 16 or 32 just to be safe.

File: src/mbgl/util/tile_cover.cpp [hunk 4]
```diff
@@ -169,25 +169,100 @@ std::vector<UnwrappedTileID> tileCover(const TransformState& state, int32_t z) {
...
+        bounds = LatLngBounds::world();
+    }
+
+    auto sw = Projection::project(bounds.southwest(), z);
+    auto ne = Projection::project(bounds.northeast(), z);
+    auto se = Projection::project(bounds.southeast(), z);
+    auto nw = Projection::project(bounds.northwest(), z);
+
+    Polygon<double> p({ {sw, nw, ne, se, sw} });
+    impl = new TileCoverImpl(z, p, false);
```
Comment (ivovandongen) [inline src/mbgl/util/tile_cover.cpp]: Don't we want to use a smart point here instead?

File: src/mbgl/util/tile_cover.hpp [hunk 5]
```diff
@@ -3,6 +3,7 @@
 #include <mbgl/tile/tile_id.hpp>
 #include <mbgl/style/types.hpp>
 #include <mbgl/util/tile_coordinate.hpp>
```
Comment (ivovandongen) [inline src/mbgl/util/tile_cover.hpp]: This header is not used here atm right? (only non-public header included here)

**Comment by asheemmamoowala** [comment]
@ivovandongen What is the use case for `OfflineRegionDefinition::tileCover` to be public? 
The `tileCover` method is marked private currently. 

If there is no public use case for the `tileCover` method, it could be moved into `OfflineDownload` or some other class where it can be used internally.

**Comment by ivovandongen** [comment]
> What is the use case for OfflineRegionDefinition::tileCover to be public?

Mostly to keep things encapsulated. Since there are wishes to expose `tileCount` on the offline region definitions (#11504 / #11108), it makes sense to keep both of them in the same place.

**Comment by ivovandongen** [comment]
@asheemmamoowala I'm getting 3 test  failures:
- [Point](https://github.com/mapbox/mapbox-gl-native/pull/11447/files#diff-bc4c4890fcf39d8b9d0f3faa7823d9acR65) - Empty {}
- [MultiPoint](https://github.com/mapbox/mapbox-gl-native/pull/11447/files#diff-bc4c4890fcf39d8b9d0f3faa7823d9acR72) - Empty {}
- [MultiLineString](https://github.com/mapbox/mapbox-gl-native/pull/11447/files#diff-bc4c4890fcf39d8b9d0f3faa7823d9acR92) - Mismatch (This might be a wrong expectation on my part, but I remember changing it to a previous version and working)

See the test results [here](https://circleci.com/gh/mapbox/mapbox-gl-native/92477?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link)

File: src/mbgl/util/tile_cover.hpp [hunk 6]
```diff
@@ -13,13 +14,36 @@ class LatLngBounds;
...
+    TileCover(const LatLngBounds&, int32_t z);
+    // When project == true, projects the geometry points to tile coordinates
+    TileCover(const Geometry<double>&, int32_t z, bool project = true);
+    ~TileCover();
+
+    //Returns false when there are no more rows to cover
+    bool next();
+    //Invokes the ScanLine callback to indicate tiles coverd for the row from [x0,x1)
+    // ScanLine may be invoked with duplcaite or overlapping ranges.
+    bool getTiles(ScanLine&);
```
[the remark this answers is not in the document]
Comment (asheemmamoowala) [reply src/mbgl/util/tile_cover.hpp]: @kkaefer I considered implementing a [`ForwardIterator`](http://en.cppreference.com/w/cpp/concept/ForwardIterator), but found that `std::iterator`s are required to be `CopyConstructible` and `CopyAssignable`. Copying the `TileCover::Impl` state is non-trivial and would affect performance if the iterator was used incorrectly.

Per our chat, the streaming should be per-tile, and not per-row.

File: src/mbgl/util/tile_cover_impl.cpp [hunk 7]
```diff
@@ -0,0 +1,285 @@
...
+#include <cmath>
+#include <assert.h>
+#include <limits.h>
+#include <algorithm>
+
+namespace mbgl {
+namespace util {
+
+void start_list_on_local_minimum(point_list& points) {
+    assert(points.size() > 2);
```
[the remark this answers is not in the document]
Comment (asheemmamoowala) [reply src/mbgl/util/tile_cover_impl.cpp]: This method is only called from `build_edge_table` when the point list is for a closed geometry, but a closed geometry with only two points should not be allowed.

**Comment by asheemmamoowala** [comment]
@ivovandongen Point and Multipoint should be fixed now. The `MultiLineString` issue was a good test case - that I hadn't considered. I incorporated all three cases into `tilecover.test.cpp` as well.

**Comment by ivovandongen** [comment]
@asheemmamoowala Great. I've rebased on your changes and adapted to the streaming tile cover in #11447

Since you've included the test cases in tile cover, I removed the redundant ones on OfflineRegionDefinition and instead added tests for encode/decode.

**Pull request merged**